src/main/java/org/apache/datasketches/common/ArrayOfStringsSerDe.java

```diff
@@ -31,45 +34,101 @@
+    Objects.requireNonNull(items, "Items must not be null");
+    if (items.length == 0) { return new byte[0]; }
+    int totalBytes = 0;
+    final int numItems = items.length;
+    final byte[][] serialized2DArray = new byte[numItems][];
+    for (int i = 0; i < numItems; i++) {
+      serialized2DArray[i] = items[i].getBytes(StandardCharsets.UTF_8);
+      totalBytes += serialized2DArray[i].length + Integer.BYTES;
     }
-    final byte[] bytes = new byte[length];
-    final WritableMemory mem = WritableMemory.writableWrap(bytes);
-    long offsetBytes = 0;
-    for (int i = 0; i < items.length; i++) {
-      mem.putInt(offsetBytes, itemsBytes[i].length);
-      offsetBytes += Integer.BYTES;
-      mem.putByteArray(offsetBytes, itemsBytes[i], 0, itemsBytes[i].length);
-      offsetBytes += itemsBytes[i].length;
+    final byte[] bytesOut = new byte[totalBytes];
+    int offset = 0;
+    for (int i = 0; i < numItems; i++) {
+      final int utf8len = serialized2DArray[i].length;
+      putIntLE(bytesOut, offset, utf8len);
+      offset += Integer.BYTES;
+      copyBytes(serialized2DArray[i], 0, bytesOut, offset, utf8len);
+      offset += utf8len;
     }
-    return bytes;
+    return bytesOut;
   }
 
   @Override
+  @Deprecated
   public String[] deserializeFromMemory(final Memory mem, final int numItems) {
+    return deserializeFromMemory(mem, 0, numItems);
+  }
+
+  @Override
+  public String[] deserializeFromMemory(final Memory mem, final long offsetBytes, final int numItems) {
+    Objects.requireNonNull(mem, "Memory must not be null");
+    if (numItems <= 0) { return new String[0]; }
     final String[] array = new String[numItems];
-    long offsetBytes = 0;
+    long offset = offsetBytes;
     for (int i = 0; i < numItems; i++) {
-      Util.checkBounds(offsetBytes, Integer.BYTES, mem.getCapacity());
-      final int strLength = mem.getInt(offsetBytes);
-      offsetBytes += Integer.BYTES;
-      final byte[] bytes = new byte[strLength];
-      Util.checkBounds(offsetBytes, strLength, mem.getCapacity());
-      mem.getByteArray(offsetBytes, bytes, 0, strLength);
-      offsetBytes += strLength;
-      array[i] = new String(bytes, StandardCharsets.UTF_8);
+      Util.checkBounds(offset, Integer.BYTES, mem.getCapacity());
+      final int strLength = mem.getInt(offset);
+      offset += Integer.BYTES;
+      final byte[] utf8Bytes = new byte[strLength];
+      Util.checkBounds(offset, strLength, mem.getCapacity());
+      mem.getByteArray(offset, utf8Bytes, 0, strLength);
+      offset += strLength;
+      array[i] = new String(utf8Bytes, StandardCharsets.UTF_8);
     }
     return array;
   }
 
+  @Override
+  public int sizeOf(final String item) {
+    Objects.requireNonNull(item, "Item must not be null");
+    if (item.isEmpty()) { return Integer.BYTES; }
+    return item.getBytes(StandardCharsets.UTF_8).length + Integer.BYTES;
+  }
+
+  @Override
+  public int sizeOf(final Memory mem, final long offsetBytes, final int numItems) {
```

Comment:
what is the purpose of this? seems like dry-run deserialization

Reply:
We do not know whether T is variable in size or not, only the user knows.  If T happens to be fixed in size, the computation of the size of the array of items is a trivial multiply, which is far less expensive than a deserialization.  If T varies in size than the user will have to scan the in-memory array and determine its size.  Not having this method would mean forcing a deserialization and then grabbing the size of the array even when a trivial multiply would do.